barretenberg/cpp/src/barretenberg/ecc/fields/field_conversion.hpp

```diff
@@ -68,75 +43,40 @@ template <typename T> constexpr size_t calc_num_bn254_frs()
-
-template <size_t N>
-inline std::array<bb::fr, N> convert_from_bn254_frs(std::span<const bb::fr> fr_vec, std::array<bb::fr, N>* /*unused*/)
-{
-    std::array<bb::fr, N> val;
-    for (size_t i = 0; i < N; ++i) {
-        val[i] = fr_vec[i];
-    }
-    return val;
-}
-
-template <size_t N>
-inline std::array<grumpkin::fr, N> convert_from_bn254_frs(std::span<const bb::fr> fr_vec,
-                                                          std::array<grumpkin::fr, N>* /*unused*/)
-{
-    std::array<grumpkin::fr, N> val;
-    for (size_t i = 0; i < N; ++i) {
-        std::vector<bb::fr> fr_vec_tmp{ fr_vec[2 * i],
-                                        fr_vec[2 * i + 1] }; // each pair of consecutive elements is a grumpkin::fr
-        val[i] = convert_from_bn254_frs<grumpkin::fr>(fr_vec_tmp);
-    }
-    return val;
-}
-
-template <size_t N>
-inline Univariate<bb::fr, N> convert_from_bn254_frs(std::span<const bb::fr> fr_vec, Univariate<bb::fr, N>* /*unused*/)
-{
-    Univariate<bb::fr, N> val;
-    for (size_t i = 0; i < N; ++i) {
-        val.evaluations[i] = fr_vec[i];
-    }
-    return val;
-}
-
-template <size_t N>
-inline Univariate<grumpkin::fr, N> convert_from_bn254_frs(std::span<const bb::fr> fr_vec,
-                                                          Univariate<grumpkin::fr, N>* /*unused*/)
+template <typename T> T convert_from_bn254_frs(std::span<const bb::fr> fr_vec)
 {
-    Univariate<grumpkin::fr, N> val;
-    for (size_t i = 0; i < N; ++i) {
-        std::vector<bb::fr> fr_vec_tmp{ fr_vec[2 * i], fr_vec[2 * i + 1] };
-        val.evaluations[i] = convert_from_bn254_frs<grumpkin::fr>(fr_vec_tmp);
+    if constexpr (IsAnyOf<T, bool>) {
+        ASSERT(fr_vec.size() == 1);
+        return fr_vec[0] != 0;
+    } else if constexpr (IsAnyOf<T, uint32_t, bb::fr>) {
+        ASSERT(fr_vec.size() == 1);
+        return static_cast<T>(fr_vec[0]);
+    } else if constexpr (IsAnyOf<T, grumpkin::fr>) {
+        ASSERT(fr_vec.size() == 2);
+        return convert_grumpkin_fr_from_bn254_frs(fr_vec);
+    } else if constexpr (IsAnyOf<T, curve::BN254::AffineElement, curve::Grumpkin::AffineElement>) {
+        using BaseField = typename T::Fq;
+        constexpr size_t BaseFieldScalarSize = calc_num_bn254_frs<BaseField>();
+        ASSERT(fr_vec.size() == 2 * BaseFieldScalarSize);
+        T val;
+        val.x = convert_from_bn254_frs<BaseField>(fr_vec.subspan(0, BaseFieldScalarSize));
+        val.y = convert_from_bn254_frs<BaseField>(fr_vec.subspan(BaseFieldScalarSize, BaseFieldScalarSize));
+        return val;
+    } else {
+        // Array or Univariate
+        T val;
+        constexpr size_t FieldScalarSize = calc_num_bn254_frs<typename T::value_type>();
+        ASSERT(fr_vec.size() == FieldScalarSize * std::tuple_size<T>::value);
+        size_t i = 0;
+        for (auto& x : val) {
+            x = convert_from_bn254_frs<typename T::value_type>(fr_vec.subspan(FieldScalarSize * i, FieldScalarSize));
```

Comment:
This feels a little funny, shouldn't the second subspan argument have an `i+1`?

Reply:
subspan takes in an offset and a count, not a start and end.